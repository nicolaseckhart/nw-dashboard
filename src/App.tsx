import * as React from 'react';
import io from 'socket.io-client';
import { Container, Nav, Navbar } from 'react-bootstrap';
import SensorData from './models/SensorData';
import WebcamData from './models/WebcamData';
import MiningData from './models/MiningData';
import { JsonMiningDump, JsonSensorDump, JsonWebcamDump } from './shared';
import { BrowserRouter as Router, Switch, Route, Link } from 'react-router-dom';
import { Dashboard } from './components/Dashboard/Dashboard';
import { SensorHistory } from './components/SensorHistory/SensorHistory';
import { WebcamOverview } from './components/WebcamOverview/WebcamOverview';
import { NotFound } from './components/NotFound/NotFound';
import { MiningOverview } from './components/MiningOverview/MiningOverview';
import WeatherSummary from './components/WeatherSummary/WeatherSummary';
import SocketDownMessage from './components/SocketDownMessage/SocketDownMessage';

interface State {
  sensorData: SensorData | null;
  webcamData: WebcamData | null;
  miningData: MiningData | null;
}

export class App extends React.Component<{}, State> {
  socket: SocketIOClient.Socket;
  disconnected = true;
  socketDisconnected = false;

  constructor(props: {}) {
    super(props);
    this.state = { sensorData: null, webcamData: null, miningData: null };
    this.socket = this.createSocket();
  }

  componentDidMount() {
    // Load debug data if in debug mode. Use this for development if the socket is down.
    if (process.env.REACT_APP_DEBUG) {
      this.setState({ sensorData: new SensorData(), webcamData: new WebcamData() });
    }

    this.connectToSocket();
  }

  componentWillUnmount() {
    this.socket.close();
  }

  createSocket = (): SocketIOClient.Socket => io(process.env.REACT_APP_WS_HOST as string);

  connectToSocket = (): void => {
    this.socket.on('connect', () => {
      this.socket.emit('authentication', {
        username: process.env.REACT_APP_SOCKET_USER,
        password: process.env.REACT_APP_SOCKET_PASSWORD,
      });
      this.socket.on('authenticated', () => {
        this.socket.on('nwmon', (data: JsonSensorDump) => {
          this.setState({ sensorData: new SensorData(data) });
          this.disconnected = false;
          this.socketDisconnected = false;
        });
        this.socket.on('nwcam', (data: JsonWebcamDump) => {
          this.setState({ webcamData: new WebcamData(data) });
          this.disconnected = false;
          this.socketDisconnected = false;
        });
        this.socket.on('mining-stats', (data: JsonMiningDump) => {
          this.setState({ miningData: new MiningData(data) });
        });
      });
    });

    setTimeout(() => {
      if (this.disconnected) {
        this.socketDisconnected = true;
      }
    }, 4000);
  };

  render = () => (
    <Router>
      <div>
        <Navbar collapseOnSelect expand="lg" variant="dark" bg="dark">
          <Link to="/">
            <Navbar.Brand>
              NW Dashboard
              <span className="jam jam-leaf ml-2 mr-3" />
            </Navbar.Brand>
          </Link>
          <Navbar.Toggle aria-controls="navigation" />
          <Navbar.Collapse id="navigation">
            <Nav className="mr-auto">
              <Link to="/mining">
                <Nav.Item className="mr-5">
                  Mining
                  <span className="jam jam-coin ml-2" />
                </Nav.Item>
              </Link>
              <Link to="/history/intakecoolant/day">
                <Nav.Item className="mr-5">
                  History
                  <span className="jam jam-history ml-2" />
                </Nav.Item>
              </Link>
              <Link to="/webcams">
                <Nav.Item className="mr-5">
                  Webcams
                  <span className="jam jam-camera-alt ml-2" />
                </Nav.Item>
              </Link>
            </Nav>
          </Navbar.Collapse>
        </Navbar>

        <Container>
          {this.socketDisconnected && <SocketDownMessage />}
          <Switch>
            <Route exact path="/">
              <Dashboard sensorData={this.state.sensorData} />
              <WeatherSummary />
            </Route>
            <Route exact path="/mining">
              <MiningOverview miningData={this.state.miningData} />
            </Route>
            <Route exact path="/history/:sensor/:scope" component={SensorHistory} />
            <Route exact path="/webcams">
              <WebcamOverview webcamData={this.state.webcamData} />
            </Route>
            <Route path="*" component={NotFound} />
          </Switch>
        </Container>
      </div>
    </Router>
  );
}

export default App;
