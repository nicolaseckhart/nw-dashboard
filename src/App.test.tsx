import * as React from 'react';
import App from './App';
import { shallow, ShallowWrapper } from 'enzyme';

function app(): ShallowWrapper {
  return shallow(<App />);
}

describe('App', () => {
  let wrapper: ShallowWrapper, instance: App;

  beforeEach(() => {
    wrapper = app();
    instance = wrapper.instance() as App;
  });

  afterEach(() => {
    instance.socket.close();
  });

  it('renders the app correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
