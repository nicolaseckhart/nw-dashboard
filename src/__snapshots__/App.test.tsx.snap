// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders the app correctly 1`] = `
<BrowserRouter>
  <div>
    <Navbar
      bg="dark"
      collapseOnSelect={false}
      expand={true}
      variant="light"
    >
      <Link
        to="/"
      >
        <NavbarBrand>
          NW Dashboard
          <span
            className="jam jam-leaf ml-2 mr-3"
          />
        </NavbarBrand>
      </Link>
      <NavbarToggle
        aria-controls="navigation"
        label="Toggle navigation"
      />
      <NavbarCollapse
        id="navigation"
      >
        <Nav
          className="mr-auto"
          fill={false}
          justify={false}
        >
          <Link
            to="/history/intakecoolant/day"
          >
            <NavItem
              className="mr-5"
            >
              History
              <span
                className="jam jam-history ml-2"
              />
            </NavItem>
          </Link>
          <Link
            to="/webcams"
          >
            <NavItem
              className="mr-5"
            >
              Webcams
              <span
                className="jam jam-camera-alt ml-2"
              />
            </NavItem>
          </Link>
        </Nav>
      </NavbarCollapse>
    </Navbar>
    <Container
      fluid={false}
    >
      <Switch>
        <Route
          exact={true}
          path="/"
        >
          <Component
            sensorData={null}
          />
        </Route>
        <Route
          component={[Function]}
          exact={true}
          path="/history/:sensor/:scope"
        />
        <Route
          exact={true}
          path="/webcams"
        >
          <Component
            webcamData={null}
          />
        </Route>
        <Route
          component={[Function]}
          path="*"
        />
      </Switch>
    </Container>
  </div>
</BrowserRouter>
`;
